### ISSUE TITLE
Type inference fails for simple case

### ISSUE THREAD

**Issue by millenialSpirou** [microsoft/pyright#2722]
Type inference fails with literal arguments known at """compile time""".
In both the following cases I get an error reporting a "GeneralTypeIssue, dummy is not iterable":

```python
class dummy:
    pass

def test(frmt="something"):
    if frmt == "something":
        return "a", "b"
    else:
        return dummy()

 a, b = test()
```

```python
class dummy
	pass

def test(frmt):
	if frmt == "something":
		return "a", "b"
	else:
		return dummy()

a, b = test("something")
```
I'd expect pyright to be able to infer statically that the return from both these calls on test is Tuple[str, str] and not Union[dummy, Tuple[str,str]]

* Using pyright with nvim+coc (coc-pyright)

**Comment by erictraut** [comment]
Return type inference for a function uses the return expressions within a function. In this example, pyright examines the two return statements in `test` and infers that it's (implied) return type annotation should be `tuple | dummy`. Pyright doesn't actually execute the code or perform cross-function analysis to further narrow the type.

If you want this to type check correctly, you would need to add type annotations with an overload.

```python
@overload
def test(frmt: Literal["something"] = ...) -> tuple[str, str]:
    ...

@overload
def test(frmt: str) -> dummy:
    ...


def test(frmt: str = "something") -> tuple[str, str] | dummy:
    if frmt == "something":
        return "a", "b"
    else:
        return dummy()
```
